### ISSUE TITLE
add custom column in the sweep_table before adding sweeps

### ISSUE THREAD

**Issue by lvsltz** [NeurodataWithoutBorders/pynwb#1071]
I propose this feature following a discussion with @t-b on the icephys channel on slack. I think this feature can be useful to group sweeps in various categories.

**Context:**
The sweep_table allows adding custom columns with add_column, but only **after** the sweeps have been added.
The trials table instead allows defining the column beforehand with add_trial_column and then the new column can be used as an argument in add_trial.

It would be nice to have a function similar to add_trial_column for the sweep_table.

**Minimal working example**:
```
from pynwb import NWBFile
from datetime import datetime
from dateutil.tz import tzlocal, UTC
from pynwb.icephys import CurrentClampStimulusSeries

nwbfile = NWBFile('a','b',datetime.now(tzlocal()))


nwbfile.add_trial_column("newcolumn","newcolumn description")

nwbfile.add_trial(0.0,1.0,newcolumn='abc')
```
```
nwbfile.trials.to_dataframe()
```
```
device = nwbfile.create_device(name='Heka ITC-1600')
elec = nwbfile.create_ic_electrode(name="elec1",
                                   description='a mock intracellular electrode',
                                   device=device)

stim = CurrentClampStimulusSeries('index0',[0.0,1.0],'A',elec,1.0,rate=1.0,sweep_number=2)
nwbfile.add_stimulus(stim)
```
```
nwbfile.sweep_table.to_dataframe()
```
```
nwbfile.sweep_table.add_column('new_sweep_column','abc',[33.0])
```
In the last line, the last argument should ideally be defined already when CurrentClampStimulusSeries is added to the nwbfile, along sweep_number.

## Checklist

- [x] Have you ensured the feature or change was not already [reported](https://github.com/NeurodataWithoutBorders/pynwb/issues)?
- [x] Have you included a brief and descriptive title?
- [x] Have you included a clear description of the problem you are trying to solve?
- [x] Have you included a minimal code snippet that reproduces the issue you are encountering?
- [x] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?

**Comment by rly** [comment]
Hi @lvsltz , thanks for the issue ticket. 
> In the last line, the last argument should ideally be defined already when CurrentClampStimulusSeries is added to the nwbfile, along sweep_number.

I think I understand, but could you please clarify this last point? Are you looking for the following syntax?

```python
nwbfile.sweep_table.add_column('new_sweep_column','abc',)
stim = CurrentClampStimulusSeries('index0',[0.0,1.0],'A',elec,1.0,rate=1.0,sweep_number=2, new_sweep_column=33)
nwbfile.add_stimulus(stim)
```

where after a new column X is defined on the sweep table, you can create a new PatchClampSeries with the argument X=x?

**Comment by lvsltz** [comment]
Yes, that's exactly what I meant.

**Comment by rly** [comment]
Thanks. On the implementation side, that case is not analogous to adding rows with custom columns to a DynamicTable and so it would be tricky to implement. The more analogous example would be something like:
```python
nwbfile.sweep_table.add_column('new_sweep_column','abc',)
stim = CurrentClampStimulusSeries('index0',[0.0,1.0],'A',elec,1.0,rate=1.0,sweep_number=2)
nwbfile.add_stimulus(stim, new_sweep_column=33)
```

Would this work as well?

The difference is that the `CurrentClampStimulusSeries` object does not store the value of that new column; the value is associated with the object only via the `SweepTable`.

**Comment by lvsltz** [comment]
I see. I think what you proposed can work very well. It's still way better than adding the additional column in one go after all sweeps. 

(Please note though that this feature will be useful only after some guidelines on how to use the `SweepTable` become available, to see for example what features can be saved in the additional columns. I have started a discussion on this with @oruebel.)

**Comment by lvsltz** [comment]
The example code in my first post above doesn't seem to work anymore with the latest release (1.1.0): add_trial_column requires a non-empty data argument. Perhaps then the requested feature doesn't make sense anymore.

```
from pynwb import NWBFile
from datetime import datetime
from dateutil.tz import tzlocal, UTC
from pynwb.icephys import CurrentClampStimulusSeries

nwbfile = NWBFile('a','b',datetime.now(tzlocal()))

nwbfile.add_trial_column("newcolumn","newcolumn description")
```
Error:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-06a6304fbed1> in <module>
      7 
      8 
----> 9 nwbfile.add_trial_column("newcolumn","newcolumn description")

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/hdmf/utils.py in func_call(*args, **kwargs)
    436                         raise_from(ExceptionType(msg), None)
    437 
--> 438                 return func(self, **parsed['args'])
    439         else:
    440             def func_call(*args, **kwargs):

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/pynwb/file.py in add_trial_column(self, **kwargs)
    525         See :py:meth:`~pynwb.core.DynamicTable.add_column` for more details
    526         """
--> 527         self.__check_trials()
    528         call_docval_func(self.trials.add_column, kwargs)
    529 

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/pynwb/file.py in __check_trials(self)
    517     def __check_trials(self):
    518         if self.trials is None:
--> 519             self.trials = TimeIntervals('trials', 'experimental trials')
    520 
    521     @docval(*get_docval(DynamicTable.add_column))

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/hdmf/utils.py in func_call(*args, **kwargs)
    436                         raise_from(ExceptionType(msg), None)
    437 
--> 438                 return func(self, **parsed['args'])
    439         else:
    440             def func_call(*args, **kwargs):

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/pynwb/epoch.py in __init__(self, **kwargs)
     29             *get_docval(DynamicTable.__init__, 'id', 'columns', 'colnames'))
     30     def __init__(self, **kwargs):
---> 31         call_docval_func(super(TimeIntervals, self).__init__, kwargs)
     32 
     33     @docval({'name': 'start_time', 'type': 'float', 'doc': 'Start time of epoch, in seconds'},

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/hdmf/utils.py in call_docval_func(func, kwargs)
    325 def call_docval_func(func, kwargs):
    326     fargs, fkwargs = fmt_docval_args(func, kwargs)
--> 327     return func(*fargs, **fkwargs)
    328 
    329 

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/hdmf/utils.py in func_call(*args, **kwargs)
    436                         raise_from(ExceptionType(msg), None)
    437 
--> 438                 return func(self, **parsed['args'])
    439         else:
    440             def func_call(*args, **kwargs):

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/pynwb/core.py in __init__(self, **kwargs)
    940                 id = ElementIdentifiers('id', data=id)
    941         else:
--> 942             id = ElementIdentifiers('id')
    943 
    944         if columns is not None:

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/hdmf/utils.py in func_call(*args, **kwargs)
    436                         raise_from(ExceptionType(msg), None)
    437 
--> 438                 return func(self, **parsed['args'])
    439         else:
    440             def func_call(*args, **kwargs):

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/pynwb/core.py in __init__(self, **kwargs)
    399              'default': list()})
    400     def __init__(self, **kwargs):
--> 401         call_docval_func(super(ElementIdentifiers, self).__init__, kwargs)
    402 
    403 

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/hdmf/utils.py in call_docval_func(func, kwargs)
    325 def call_docval_func(func, kwargs):
    326     fargs, fkwargs = fmt_docval_args(func, kwargs)
--> 327     return func(*fargs, **fkwargs)
    328 
    329 

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/hdmf/utils.py in func_call(*args, **kwargs)
    436                         raise_from(ExceptionType(msg), None)
    437 
--> 438                 return func(self, **parsed['args'])
    439         else:
    440             def func_call(*args, **kwargs):

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/pynwb/core.py in __init__(self, **kwargs)
    279             {'name': 'data', 'type': ('array_data', 'data', Data), 'doc': 'the source of the data'})
    280     def __init__(self, **kwargs):
--> 281         call_docval_func(super(NWBData, self).__init__, kwargs)
    282         self.__data = getargs('data', kwargs)
    283 

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/hdmf/utils.py in call_docval_func(func, kwargs)
    325 def call_docval_func(func, kwargs):
    326     fargs, fkwargs = fmt_docval_args(func, kwargs)
--> 327     return func(*fargs, **fkwargs)
    328 
    329 

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/hdmf/utils.py in func_call(*args, **kwargs)
    436                         raise_from(ExceptionType(msg), None)
    437 
--> 438                 return func(self, **parsed['args'])
    439         else:
    440             def func_call(*args, **kwargs):

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/pynwb/core.py in __init__(self, **kwargs)
     32     @docval({'name': 'name', 'type': str, 'doc': 'the name of this container'})
     33     def __init__(self, **kwargs):
---> 34         call_docval_func(super(NWBBaseType, self).__init__, kwargs)
     35         self.__fields = dict()
     36 

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/hdmf/utils.py in call_docval_func(func, kwargs)
    325 def call_docval_func(func, kwargs):
    326     fargs, fkwargs = fmt_docval_args(func, kwargs)
--> 327     return func(*fargs, **fkwargs)
    328 
    329 

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/hdmf/utils.py in func_call(*args, **kwargs)
    434                     if parse_err:
    435                         msg = ', '.join(parse_err)
--> 436                         raise_from(ExceptionType(msg), None)
    437 
    438                 return func(self, **parsed['args'])

~/miniconda3/envs/nwb2/lib/python3.7/site-packages/six.py in raise_from(value, from_value)

TypeError: None is not allowed for 'data' (expected 'ndarray, list, tuple, Dataset, HDMFDataset or data', not None)


```

**Comment by rly** [comment]
I believe this error is due to an unfortunate issue where the latest version of PyNWB (1.1.0) is not compatible with the latest version of one of its dependencies, HDMF, currently at 1.3.2. I will release PyNWB 1.1.1 tomorrow to resolve this, hopefully.

**Comment by rly** [comment]
@lvsltz Please try again with PyNWB 1.1.1 and let me know how it goes. Thanks!

**Comment by lvsltz** [comment]
It's working as before now, thanks!
